### PR TITLE
Uwlidar

### PR DESCRIPTION
**Pull request by M1chaelM**
This is a very basic standalone demo with the new 3d at depth Lidar Model, based on the specs listed here: https://www.3datdepth.com/product/subsea-lidar-sl3

It is not quite ready but think it would especially benefit from a sanity check on a few things I've been puzzling about:

- The specs say this produces a 30x30 section that is 1450 x 1450 points (if I'm reading it right?). That is a lot higher than a velodyne, and if I try to get Gazebo to give me anywhere near 1450 horizontal and vertical beams it hangs. I can get up to about 100th of this resolution. (145 x 145) comfortably.
- Specs also list this is a 20.9Kg cylinder with a height of 355 mm and a diameter of 210 mm. That seems very dense, and gives me some large moments of inertia: 0.277 for xx and yy, and 0.115 for zz. So again wondering if I'm reading that right.

Lastly, I'm assuming we want this to publish point cloud data to a topic, but that is not done yet.

To test, you can run:
`
roslaunch nps_uw_sensors_gazebo uw_lidar_standalone.launch
`

Many thanks for looking this over!

### CODE REVIEW

**Comment by M1chaelM** [comment]
I forgot to add:

Does anyone know what the <resolution> element in the **<range>** element does? The documentation says "linear resolution of each Lidar ray." I'm assuming a "ray" refers to the whole ray element, so I thought this might be the depth measurement resolution, but I can't find any information on it, and it doesn't seem to make any difference what value is in the field--100,10,.001, and even 0 all give the same visualization.

**Comment by BruceMty** [comment]
Gazebo 9 file GpuRaySensor.hh says this about range resolution:

      /// \brief Get the range resolution
      ///      If RangeResolution is 1, the number of simulated rays is equal
      ///      to the number of returned range readings. If it's less than 1,
      ///      fewer simulated rays than actual returned range readings are
      ///      used, the results are interpolated from two nearest neighbors,
      ///      and vice versa.
      /// \return The Range Resolution
      public: double RangeResolution() const;

I don't think it is used in Gazebo or ROS except in Gazebo tests.

**Comment by M1chaelM** [comment]
Thanks for finding that! I'm a bit confused about when I should be using real (simulated) beams and when I should be using interpolated simulated beams...

But if the answer is just that I can get a 1450 x 1450 point cloud using 145 x 145 with resolution of 0.1, that would be great.

**Comment by BruceMty** [comment]
That is correct.  If you want a different number of data points than the number of samples the lidar takes during a sweep then use resolution to interpolate or average.  If your resolution is so large it slows down Gazebo then use a smaller resolution and indicate that it does not match the product you are simulating.

**Comment by M1chaelM** [comment]
This is about ready to go now, but could definitely benefit from some review. 

- To test, first make sure you have installed the plugins:
`sudo apt install ros-melodic-velodyne-gazebo-plugins`
- Then run
  `roslaunch nps_uw_sensors_gazebo uw_lidar_standalone.launch`
- Unpause the simulation. Lidar should sweep through the tank.
- `rostopic list` should show the following topics:
`/nps_gazebo_ros_pulse_lidar_standalone/lidar_pan_cmd
/nps_gazebo_ros_pulse_lidar_standalone/lidar_tilt_cmd
/nps_gazebo_ros_uw/pulse_lidar/points
`
- To control the lidar stand, you can run:
` rostopic pub /nps_gazebo_ros_pulse_lidar_standalone/lidar_pan_cmd std_msgs/Float32 -- <angle>`
Where <angle> is an angle in radians that you want the device to rotate to.
- Likewise you can do the same with 
` rostopic pub /nps_gazebo_ros_pulse_lidar_standalone/lidar_tilt_cmd std_msgs/Float32 -- <angle>`
- You'll notice tilt doesn't work as well. I think this is because of gravity of some physics I haven't figured out.
- Generally, I don't think I have the PIDs tuned correctly. Even the pan eventually oscillates. Any help / suggestions would be much appreciated.

**Comment by bsb808** [comment]
I get the following error:
```
-- Configuring done
CMake Error at nps_uw_sensors_gazebo/CMakeLists.txt:34 (add_executable):
  Cannot find source file:

    src/lidar_test.cc

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx


CMake Error at nps_uw_sensors_gazebo/CMakeLists.txt:23 (add_library):
  Cannot find source file:

    src/gazebo_ros_pulse_lidar.cpp

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx

```

Perhaps some of the source files needed to be added to git?

**Comment by bsb808** [comment]
I know Bruce is signed up to provide the review, but one general suggestion.  If there are details that need to be addressed (e.g., improved tuning of the PID control) you might want to have the PR be the first step and add issues to describe the next steps.

Also, it would be nice to have an RVIZ configuration as part of the example launch file to visualize the point clouds.

File: CMakeLists.txt [hunk 1]
```diff
@@ -4,24 +4,40 @@ project(nps_uw_sensors_gazebo)
...
 
 catkin_package()
 
 ## Plugins
-add_library(nps_gazebo_ros_gpu_sonar_single_beam_plugin
-            src/nps_gazebo_ros_gpu_sonar_single_beam.cpp)
-target_link_libraries(nps_gazebo_ros_gpu_sonar_single_beam_plugin
-                      GpuRayPlugin ${catkin_LIBRARIES})
+add_library(gazebo_ros_pulse_lidar_plugin src/gazebo_ros_pulse_lidar.cpp)
+add_library(nps_gazebo_ros_gpu_sonar_single_beam_plugin src/nps_gazebo_ros_gpu_sonar_single_beam.cpp)
```
Comment (BruceMty) [inline CMakeLists.txt]: Can we keep the add_library(nps_gazebo_ros_gpu_sonar_single_beam_plugin as two lines?  I don't want them merged, merging implies something changed, and being split is acceptable with existing convention.  Also, I try to keep line length to 80 where convenient.

File: CMakeLists.txt [hunk 2]
```diff
@@ -4,24 +4,40 @@ project(nps_uw_sensors_gazebo)
...
 # for launch
 install(DIRECTORY launch
         DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION}
         PATTERN "*~" EXCLUDE)
 
+# Build the stand-alone test program
+add_executable(lidar_test src/lidar_test.cc)
+
+if (${gazebo_VERSION_MAJOR} LESS 6)
+  # These two
```
Comment (BruceMty) [inline CMakeLists.txt]: Lets not add any new code that supports Gazebo version less than 9.  I don't mind leaving alone existing C++ ifdefs.  Wiki indicates we support Gazebo 9+ for ROS Melodic.

File: CMakeLists.txt [hunk 3]
```diff
@@ -4,24 +4,40 @@ project(nps_uw_sensors_gazebo)
...
+                      GpuRayPlugin ${catkin_LIBRARIES} ${GAZEBO_LIBRARIES} ${roscpp_LIBRARIES})
 
 # for launch
 install(DIRECTORY launch
         DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION}
         PATTERN "*~" EXCLUDE)
 
+# Build the stand-alone test program
+add_executable(lidar_test src/lidar_test.cc)
+
```
Comment (BruceMty) [inline CMakeLists.txt]: I don't seem to have src/lidar_test.cc.  Please check.

Comment (M1chaelM) [reply CMakeLists.txt]: That reference is outdated so I removed it. Thanks!

Comment (BruceMty) [reply CMakeLists.txt]: Great!

File: CMakeLists.txt [hunk 4]
```diff
@@ -4,21 +4,26 @@ project(nps_uw_sensors_gazebo)
...
 
 catkin_package()
 
 ## Plugins
-add_library(nps_gazebo_ros_gpu_sonar_single_beam_plugin
-            src/nps_gazebo_ros_gpu_sonar_single_beam.cpp)
-target_link_libraries(nps_gazebo_ros_gpu_sonar_single_beam_plugin
-                      GpuRayPlugin ${catkin_LIBRARIES})
+add_library(gazebo_ros_pulse_lidar_plugin src/gazebo_ros_pulse_lidar.cpp)
+add_library(nps_gazebo_ros_gpu_sonar_single_beam_plugin src/nps_gazebo_ros_gpu_sonar_single_beam.cpp)
```
Comment (BruceMty) [inline CMakeLists.txt]: Looks like src/nps_gazebo_ros_gpu_sonar_single_beam.cpp is not here, please check.

Comment (M1chaelM) [reply CMakeLists.txt]: Scroll down a few lines. We can keep these separate if you want. Does this compile for you?

Comment (BruceMty) [reply CMakeLists.txt]: Oops, I reported the wrong filename.  Please check for presence of src/gazebo_ros_pulse_lidar.cpp

Comment (M1chaelM) [reply CMakeLists.txt]: ugh! Sorry, it should be there now.

Comment (BruceMty) [reply CMakeLists.txt]: Fixed, good.

File: CMakeLists.txt [hunk 5]
```diff
@@ -4,21 +4,26 @@ project(nps_uw_sensors_gazebo)
...
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${GAZEBO_CXX_FLAGS}")
 
 catkin_package()
 
 ## Plugins
-add_library(nps_gazebo_ros_gpu_sonar_single_beam_plugin
-            src/nps_gazebo_ros_gpu_sonar_single_beam.cpp)
-target_link_libraries(nps_gazebo_ros_gpu_sonar_single_beam_plugin
-                      GpuRayPlugin ${catkin_LIBRARIES})
+add_library(gazebo_ros_pulse_lidar_plugin src/gazebo_ros_pulse_lidar.cpp)
```
Comment (BruceMty) [inline CMakeLists.txt:24]: Library name conflict: Here we make libgazebo_ros_pulse_lidar_plugin.so but models/3dad_sl3/model.sdf wants libgazebo_ros_velodyne_gpu_laser.so.  To test, you might remove the old .so from uuv_ws/devel/lib.

Comment (M1chaelM) [reply CMakeLists.txt:24]: Bruce, we need both of those plugins. One is from the Gazebo library, and one is what I wrote. What error are you seeing?

Comment (BruceMty) [reply CMakeLists.txt:24]: @M1chaelM , on "roslaunch nps_uw_sensors_gazebo uw_lidar_standalone.launch" Gazebo gives: "[Err] [Plugin.hh:178] Failed to load plugin libgazebo_ros_velodyne_gpu_laser.so: libgazebo_ros_velodyne_gpu_laser.so: cannot open shared object file: No such file or directory".

Comment (M1chaelM) [reply CMakeLists.txt:24]: Hm. I just ran catkin_make clean then catkin_make and launched it again with no error. Sorry to ask but can you check that you sourced the setup.bash?

Comment (M1chaelM) [reply CMakeLists.txt:24]: The plugin it is looking for should be at /opt/ros/melodic/lib/libgazebo_ros_velodyne_gpu_laser.so
Do you have it?

Comment (BruceMty) [reply CMakeLists.txt:24]: Yes, I sourced setup.bash.  No, I don't have it.  I see, I need to install velodyne_gazebo_plugins...

**Comment by M1chaelM** [comment]
@bsb808 Thanks.

- I added an rviz configuration.
- I was hoping the PID fix would be trivial for someone who understands PIDs, but maybe that's not the case? In that case I can make an issue for it.

File: CMakeLists.txt (hunk 5 above)
Comment (BruceMty) [reply CMakeLists.txt:24]: Is the installation for this plugin on the Wiki?  I don't see it at https://github.com/Field-Robotics-Lab/DAVE/wiki/Install-Directly-on-Host

Comment (M1chaelM) [reply CMakeLists.txt:24]: Oh, good to know. I'm not sure when that got installed.

Comment (M1chaelM) [reply CMakeLists.txt:24]: Does this work:
`sudo apt install ros-melodic-velodyne-gazebo-plugins`

Comment (BruceMty) [reply CMakeLists.txt:24]: Yes, fixed.

File: CMakeLists.txt (hunk 1 above)
Comment (M1chaelM) [reply CMakeLists.txt]: OK, I cleaned up this file. Please see if it still builds in case I missed anything.

Comment (BruceMty) [reply CMakeLists.txt]: Resolved.

File: CMakeLists.txt (hunk 2 above)
Comment (BruceMty) [reply CMakeLists.txt]: Resolved.

**Comment by BruceMty** [comment]
> I get the following error:
> 
> ```
> -- Configuring done
> CMake Error at nps_uw_sensors_gazebo/CMakeLists.txt:34 (add_executable):
>   Cannot find source file:
> 
>     src/lidar_test.cc
> 
>   Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
>   .hxx .in .txx
> 
> 
> CMake Error at nps_uw_sensors_gazebo/CMakeLists.txt:23 (add_library):
>   Cannot find source file:
> 
>     src/gazebo_ros_pulse_lidar.cpp
> 
>   Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
>   .hxx .in .txx
> ```
> 
> Perhaps some of the source files needed to be added to git?

Resolved.

**Comment by BruceMty** [comment]
I'm fine with this PR, nice work.

Remaining comments:

* This does not return a point cloud, I'm not sure if it needs to in this release.  When it does, we will need to comment that this device will return a point cloud of 145X145 instead of 1450X1450 because the larger size burdens the simulator.
* Please update https://github.com/Field-Robotics-Lab/dave/wiki/Install-Directly-on-Host to indicate need for "sudo apt install ros-melodic-velodyne-gazebo-plugins".
* I did not evaluate rviz.  Doesn't rviz need point cloud?

**Comment by M1chaelM** [comment]
Bruce,
I get a point cloud, so something is amiss. It sounds like the velodyne plugin is still not working for you. When you run the launch file do you see any Gazebo messages referencing it? You should see:
`[ INFO] [1591399557.181100550]: Velodyne GPU laser plugin ready, 145 lasers`

**Comment by M1chaelM** [comment]
> That is correct. If you want a different number of data points than the number of samples the lidar takes during a sweep then use resolution to interpolate or average. If your resolution is so large it slows down Gazebo then use a smaller resolution and indicate that it does not match the product you are simulating.

I interpreted the statement above to mean that we could still get 1450x1450 using 145x145 with 0.1 resolution. Was that incorrect?

**Comment by BruceMty** [comment]
> > That is correct. If you want a different number of data points than the number of samples the lidar takes during a sweep then use resolution to interpolate or average. If your resolution is so large it slows down Gazebo then use a smaller resolution and indicate that it does not match the product you are simulating.
> 
> I interpreted the statement above to mean that we could still get 1450x1450 using 145x145 with 0.1 resolution. Was that incorrect?

Thanks for asking.  I believe that is correct.  If you get the chance, try it and see if performance is acceptable.  Internally, 145X145 points will be calculated, then a 1450X1450 point cloud should be extrapolated and returned.

**Comment by M1chaelM** [comment]
OK I submitted the outstanding issues as issues per Brian's suggestion and I will now merge this.

**Comment by BruceMty** [comment]
> Bruce,
> I get a point cloud, so something is amiss. It sounds like the velodyne plugin is still not working for you. When you run the launch file do you see any Gazebo messages referencing it? You should see:
> `[ INFO] [1591399557.181100550]: Velodyne GPU laser plugin ready, 145 lasers`

Excellent, I see this.  Additionally, after unpausing, "rostopic type /nps_gazebo_ros_uw/pulse_lidar/points" shows "sensor_msgs/PointCloud2" and "rostopic echo /nps_gazebo_ros_uw/pulse_lidar/points" shows the point cloud.

**Pull request merged**